### DISCUSSION TITLE
(PoC 이후)디자인 시스템 결정 🤝 - 파운데이션 vs 컴포넌트

### DISCUSSION THREAD

**Discussion by peanudge** [InnerCircleA/design-system#12] [Ideas]
✍️ 저희가 **[PoC](https://github.com/InnerCircleA/design-system/discussions/6) 를 수행하면서 얻은 지식**과 **주변 디자인 시스템과 관련된 지인들의 인터뷰를 기반**으로 이번 디자인 시스템 프로젝트의 방향성을 다시 피봇팅하려합니다.

이번 결정은  PoC 때보다 충분한 경험과 자료가 모였기 때문에  더욱 합리적인 근거로 결정이 되면 좋을 것 같습니다. 

추가로 저희가 이야기해볼만한 부분은 흐름상  "파운데이션 기반 디자인 시스템"을 선택할 것 같지만 구체적으로 **어떤 부분을 핵심적으로** 도전해볼지 고민해봐야할 것 같아요. 

실무에서 `디자인 <-> 개발자 <-> 배포/운영`  워크플로우 중 문제가 발생되는 부분(원인)이 무엇인지 명확히 파악하는 과정도 함께 이뤄지면서 설계가 진행되었으면 좋겠습니다.

# 목표

- 디자인시스템 결정 (파운데이션 기반 ~~or 컴포넌트 기반~~)
- 아키텍처 설계
- 리포 및 서비스 구현

# TODO

- [x]  디자이너, 개발자 인터뷰 모아서 정리. @Jeewhan
- [x] AST로 얻을 수 있는 정보들을 모두 모아보기 @peanut-lover 
- [x] Storybook MDX 문서화에 숙련되어있기 => 개인 미션하면서 해봤습니다. @peanut-lover 
- [x] 실제 구현체가 있을 [레포](https://github.com/InnerCircleA/mono-repo)를 만들고 아키텍처를 README에 정리
 

# 기한 

타임라인은 `2022.12.26 ~ 2023.01.04` 입니다.

**Comment by Jeewhan** [comment]
# 방향성 조정 결과, 디자인 시스템 목표 수립을 위해 진행한 사용자 인터뷰 요약

저희가 만들고자 하는 디자인 시스템의 목표는 아래와 같습니다.

### 개별 컴포넌트가 프로덕트 어디 어디에서 쓰이고 있는지를 알려주고, 컴포넌트의 시각적 변화를 위한 엔지니어링 비용을 최소화하여, 디자이너가 시각적 변화에 대한 자유도를 가질 수 있도록 한다.

12월 28일 20시에 회의를 하면서 후술할 사용자 인터뷰 내용들을 공유한 뒤 위의 목표를 수립하고 아래와 같은 플로우를 그려보게 되었습니다.

![image](https://user-images.githubusercontent.com/14256139/210085108-f5ca1636-396c-4a3b-b1a6-a554232e483c.png)

원래는 각 컴포넌트 별로 쓰이고 있는 프로덕트 위치를 알려주고, 컴포넌트가 쓰이고 있는 Figma 시안과의 연결을 해주는 것을 1순위로 고려했었지만 그것이 디자이너들의 Figma 사용패턴상 어렵다고 판단하게 되어, 구현 난이도 등을 고려했을 때 2순위로 생각했었지만 사용자 인터뷰 내용상 더 효과적이라고 파악하게 된, Figma의 변화를 바탕으로 컴포넌트를 수정해주는 것에 초점을 맞추게 되었습니다.

그러기 위해 지호님(@peanut-lover) 께선 Webpack Plugin을 통해 컴포넌트의 사용처 외에도 보여줄 수 있는 데이터들을 알아보고, Nx로 Monorepo를 구현하고, Storybook으로 컴포넌트를 보여주는 등의 접근을 진행해주시고, 그 사이에 저는 Button이라는 Figma에서의 컴포넌트에서 시각적 변화가 있을 때 그것이 컴포넌트 라이브러리의 Button의 스타일 변화로 이어질 수 있게 하는 MVP를 다음 세션 전까지 구현하기로 하고 그 과정을 중간중간 공유를 드리기로 하였습니다.

---

인터뷰 대상자는 프로덕트 디자이너 세 분, 플랫폼 엔지니어 한 분이었습니다.

인터뷰를 하기 전에 가지고 있었던 궁금증들은 아래와 같습니다.
1. 정말로 디자이너가 디자인시스템을 통해 알고 싶어하는 (기존에는 몰랐던) 정보는 무엇일까? Good to have가 아닌 없어서 정말로 불편한 것이 무엇인가? 디자이너가 생각하는 이상적 플로우는 무엇일까?
2. Figma에선 Component Instance에서 Main Component로는 갈 수 있는데, Main Component에서 Component Instance로는 가지 못한다. 이것이 불편하지 않은가? (혹시 Instance Finder라는 Figma Plugin을 사용하고 있는가?) Main Component의 Instance를 모른다면 Main Component를 수정했을 때의 여파를 가늠할 수가 없다. 이 부분을 해결되면 이익이 큰가?

---

1. 디자이너 A(음성 인터뷰)

인터뷰를 하면서 디자이너별로 Figma 숙련도가 굉장히 많이 차이난다는 것을 느꼈습니다.

생산성 측면에서 당연히 사용되어야 한다고 느껴지는 Tokens나 Component 등의 Figma 기능을 쓰지 않고, 기능의 사용 여부에 대해 취향의 영역으로 이야기하는 것을 보고 솔직하게 충격을 정말 많이 받았습니다.

흔히들 말하는 '디자인 시스템은 규모가 어느 정도 있는 팀에서만 할 수 있고 해야 한다'는 인식에 대해선 어느 정도 동의할 수 있지만, Figma에서 이미 제공되는 기능들을 사용하지 않는 것에는 동의하기 어려웠습니다.

물론 디자이너의 디자인 툴 사용 능력은 문제 해결을 위해 사용하는 여러 층위에서 가장 바깥에 있고 부차적인 것일테지만, 그럼에도 아예 중요하지 않은 것은 아니라고 생각해보게 되었습니다. 게다가 디자이너에게 Figma는 엔지니어의 에디터보다는 언어에 가깝다고 느껴집니다. 그 이유는 엔지니어의 에디터는 결과물에 영향을 미치지 않지만 디자이너의 Figma는 협업을 위한 시안의 변화 대응력에 직접적으로 영향을 끼친다고 느껴서입니다. Figma 기능을 쓰지 않게 되면, 마치 엔지니어가 함수를 쓰지 않는 것과 비슷하다고 느껴졌습니다.

잘하는 엔지니어에겐 특정 언어에 대한 경험이 크게 중요하진 않겠지만, 그럼에도 어느 하나의 언어를 깊이있게 다뤄본 적이 있어야 그것을 바탕으로 스노우볼링을 굴려서 다른 언어도 잘 할 수 있을 것 같다고 생각합니다. 마찬가지로 디자이너도 툴에 대해 적어도 한 가지는 깊이있게 다루는 능력이 필요할 것 같습니다.

그리고 위에서 말한 것처럼 이미 도구를 잘 사용할 수 있는게 디자인 시스템 구축/유지/활용과도 연결될 것이라 생각되어, 디자인 시스템의 고객이자 생산자가 될 수 있는 디자이너의 전제과도 이어질 것 같습니다.

---

2. 디자이너 B(대면 인터뷰)

빠른 구현을 거듭해야 하는 상황에서 대략적으로 어떤 플로우로 작업하는지 들어볼 수 있었습니다. 디자인 팀 내에선 파운데이션과 컴포넌트가 갖춰져있어서 그것들을 가져와서 활용하고 필요시 디자인시스템팀 디자이너 분들에게 문의해서 구현하는 흐름이었습니다.

Figma에서 Main Component에서 Component Instance로는 갈 수 없다는 점에 대해서 모르고 계셨습니다. 그래서 그것을 파악해야 할 때는 어떻게 하는지 물어보니 엔지니어에게 문의한다는 답변을 받았습니다. 엔지니어 측에선 Figma와 달리 해당 컴포넌트가 어디서 쓰이고 있는지 바로 참조할 수 있다는 사실은 모르고 계셨지만, 엔지니어에게 물어보면 어디서 쓰이고 있는지를 파악할 수 있다는 점은 알고 있었고, 제 설명을 듣고 나서야 왜 엔지니어들이 금방 찾을 수 있는지를 이해하셨습니다.

또한 디자이너 A와 달리 디자인 시안에서 AS-IS는 남겨두지 않는 스타일이었습니다. 또한 자신이 담당하고 있는 페이지에 대한 최신화된 전체 시안을 따로 관리하지는 않았고 매 Feature마다 관련된 플로우에 대한 TO-BE를 작성하여 엔지니어에게 제공하는 형식이었습니다. (디자이너 D와의 인터뷰까지 해보면서, 디자이너마다 페이지 구분 등의 Figma 활용 방식이 많이 다르다는 점을 알 수 있었고, 그런 부분들을 통일해야 하는 니즈 등이 낮다는 점을 파악한 뒤 디자이너 행동의 변화를 통해서만 가능한 디자인 시스템적 요소는 최후순위로 미뤄야겠다는 판단을 내리게 되었습니다. 디자이너 D와의 인터뷰까지 읽어보시면 이해가 더 잘 되실 듯 합니다.)

자신의 일이 급한 경우가 너무나 많기 때문에 디자인 시스템에 대한 기대감은 특별히 없었지만 회사의 디자인 시스템 관련해서 어려움을 파악해보았을 때는 디자이너 측 디자인 시스템이 엔지니어 측과 연계되어있지 않아서 개별 목적조직의 엔지니어마다 따로 따로 대응해주고 있는 느낌을 말씀해주셨습니다. 그래서 자신이 시안을 넘겨준 뒤의 과정에서 뭔가 원활하고 빠르게 진행되고 있지는 않는 것으로 느끼고 계셨는데, 할 일이 너무 많아 미처 그 부분이 무엇인지까지는 자세히 파악하고 있지 못하고 있다고 말씀해주셨습니다.

---

3. 엔지니어 C(텍스트 기반 인터뷰)

현재 회사에서 디자인 시스템을 만드는 팀에 소속되어 일하고 계신 분입니다. 당연히 디자이너와는 바라보는 시각이 많이 달랐고, 지금의 디자인 시스템을 만들기 시작한지 오래 되지 않았기에 디자인 시스템의 목표와 관련해서 현재는 디자이너들에게 무엇이 필요할지보다는, 엔지니어링 측의 생산성에 집중하고 있다는 점을 알 수 있었습니다. 그래서 디자이너에 관한 인사이트보다는 사내 디자인 시스템에서 어떤 부분들을 참고하며 진행하고 있는지, 이후에 어떤 것들을 구현하고 싶어하시는지 등에 대해 들을 수 있었습니다.

---

4. 디자이너 D(화상 인터뷰)

다른 디자이너들과 나눈 이야기에 대해 간략히 설명하고 역시나 아래 질문에 대한 답을 찾고자 했습니다.
- 디자이너에게 유의미한 디자인 시스템이 되려면 무엇이 필요한가?
- 파운데이션과 컴포넌트의 사용처를 알려주면 되는가? 피그마에서 어떤 정보를 얻어서 보여주어야 의미있는가?

이 인터뷰이는 여러 가지로 디자이너의 작업 흐름에 대해 설명을 굉장히 잘 해주어, 디자이너에 대해 이전보다는 조금 더 깊이있게 이해할 수 있게 되었다고 느꼈습니다.

후술할 인터뷰 주제들과 별개로 재미있었던 포인트 중 하나는 맡고 있는 페이지 또는 프로덕트의 성격에 따라서 사용하는 컴포넌트도 굉장히 달라진다는 점이었습니다. 인터뷰이는 depth가 없는 탭을 맡고 있어서 checkbox나 radio처럼 분기를 제어하는 등의 컴포넌트는 쓸 일이 없고, 그렇기 때문에 일반적으로 디자인 시스템 컴포넌트로 되어있는 것을 활용하는 빈도가 다른 프로덕트 디자이너들에 비해 낮고 오히려 새로운 UI 요소를 만들어서 실험해보고 괜찮으면 그것이 디자인 시스템의 컴포넌트로도 반영되는 흐름으로 일하고 있다는 말을 들었습니다. 엔지니어도 맡고 있는 프로덕트과 파트마다 평소 하는 고민과 흐름이 많이 달라지듯이 디자이너는 또 다른 의미로 다양한 방식과 시점으로 일한다는 점이 어떻게 보면 당연하지만 제겐 매우 신선했습니다.

인터뷰 내용을 통해 우선 디자이너의 니즈도 플랫폼 디자이너와 프로덕트 디자이너에 따라 크게 다를 수 있다는 점을 알게 되었습니다.
중요한 전제는 프로덕트 디자이너는 최종을 구현하는 과정에서 굉장히 많은 화면을 만들게 되고 그것들을 일일이 정리하지 않으며 내가 임의로 지정한 최최최종에 대한 버전관리만 신경쓴다는 점이 굉장히 중요한 포인트였습니다. 상대적으로 엔지니어링은 하나의 현재본으로 귀결되는데 비해 Figma에선 과정작들과 최종작들이 혼재되어있고, 담당 프로덕트 디자이너는 그것을 구별할 수 있지만 Figma 자체에서는 그것들이 모두 동일한 각각의 화면을 존재할 것입니다. 일반적으로 프로덕트 디자이너들의 작업 흐름이 그렇다고 한다면, 디자인 시스템의 특정 목적을 달성하기 위해서 모두의 행동 변화를 통해서만 가능한 방식으로는 달성이 어려울 것이라 봤습니다.

Figma 에서 Component Instance 위치를 파악해서 Main Component를 수정했을 때 생길 프로덕트에서의 여파를 알고자 하는 니즈는 플랫폼 디자이너에겐 크게 있을 것으로 추측되지만, 그것을 Figma 관련 자동화로 해내기 어려울 것이라 판단하게 된 이유는 앞에서 말한 것처럼 Figma에선 너무 모든 것이 혼재되어있기 때문이었습니다. 물론 디자이너들도 draft, graveyard 등을 구분하면 좋기야 하겠지만 그렇게 잘 하지 않고, 그렇게 바꾸긴 어려울 것 같다고 인터뷰이와 인터뷰어 모두 공감대를 형성했습니다. (Figma에서 이걸 쉽게 하도록 유도하는 기능이 생긴다면 조금 달라질 수도 있을 것.)

실제로는 플랫폼 디자이너들이 특정 컴포넌트들이 쓰이고 있는 곳이 어딘지를 프로덕트 디자이너들에게 수소문하고, 그걸 또 프로덕트 디자이너들은 필요할 때는 엔지니어에게 도움을 청해서 파악하는 형태의 비효율이 반복되고 있고 그것이 생각보다 정리되기 쉬운 형태로 지금 Figma가 관리되고 있지 않다고 추측했습니다.

설령 플랫폼 디자이너가 디자인 시스템을 업데이트했을 때 프로덕트 디자이너의 시안에서 깨지더라도 이미 구현이 끝난 화면들이라면 그것들의 업데이트까진 프로덕트 디자이너가 굳이 신경쓰지 않는다고 했습니다. 또한 플랫폼 디자이너가 Team Library를 업데이트했지만 changelog 상에서 프로덕트 디자이너들에게 아직은 업데이트하지 말아달라고 하는 경우도 많고, 그것은 아직 플랫폼 디자이너들이 변경의 여파에 대해 충분히 파악하지 못한 시점인 것 같다는 말을 들었습니다.

엔지니어링에서도 구현을 위한 과정이 프로덕트 코드 내에도 살아있는 경우가 있긴 하지만, 사용자에겐 그것이 노출되지 않을 것이기도 하고, git을 활용해서 프로덕트 전체가 버전관리 대상이라고 여겨지는 것에 비해, 디자이너들은 자신이 임의로 생각하는 특정 페이지 등에 분산되어있는 특정 결과물들만을 버전관리의 대상이라고 여기는 경향이 있는 것 같아 Figma 전체 대상으로 하는 무엇의 자동화는 실효성이 많이 떨어질 것이라는 판단을 하게 되었습니다. (디자인 시스템 특정 요소에 대해선 이게 큰 장애물이 될 수도 있을 것 같습니다.)

또한 지금의 비효율 중 하나는 디자인 팀 내 디자인 시스템과 엔지니어링 측과 유기적으로 연결되어있지 않기 때문에 디자인시스템팀 플랫폼 디자이너가 목적조직 프로덕트 디자이너에게 디자인 시스템 컴포넌트의 변화를 알려주면 프로덕트 디자이너는 그것의 변화를 알고 업데이트한 뒤에 프로덕트 엔지니어에게 그걸 알려주어 업데이트를 할텐데, 이 과정을 목적조직마다 모두 하고 있을 것이라는 점이었습니다.

그래서 관련하여 혹시 컴포넌트의 기능이 아닌 모양만의 변화를 하는 경우도 많은지에 대해서 질문했더니 자신을 포함한 디자이너들에겐 그런 니즈가 많을 것 같다는 답변을 들었습니다. (기능의 변화는 어쩔 수 없이 엔지니어링적 비용도 발생할 수 밖에 없을 것이므로 모양의 변화 즉 엔지니어링 비용을 거의 들이지 않아도 될 것 같은 영역에 대한 필요성이 있는지 궁금했습니다.)

현재는 디자인 시스템이 충분한 성숙도를 갖추지 못해서 어떤 공통의 변화가 있을 때 그 변화가 반영되어야 하는 모든 화면을 담당하는 개별 엔지니어들을 모두 설득해서 그 변화를 반영시켜야 하고 그렇기 때문에 그게 잘 이루어지기도 힘들고, 그런 종류의 변화는 굉장히 우선순위가 밀릴 수 밖에 없다는 사실을 알기 때문에 디자이너들도 룩 앤 필에 대한 변화를 전혀 엄두를 내고 있지 못하다고 했습니다. 소소한 변화들이 모이면 화면의 톤 앤 매너를 바꿀 수 있지만 앞에서 말한 이유로 지금은 시각적 부채를 해소하기가 매우 힘들다고 했습니다. 지금도 플랫폼 디자이너들이 전체적인 무드에 대한 업데이트도 적극적으로 하고 싶어하기 때문에 프로덕트 디자이너들이 그런 요청을 하면 호의적일 것이므로, 구현 과정에서의 엔지니어링적 비용이 많이 드는 것만 아니라면 시각적 개선을 훨씬 더 적극적으로 많이 할 수 있을 것 같다고 말해주셨습니다.

그래서 제가 이해한 아래의 플로우를 말로 설명해주었고 대략적으로 원하는 바와 일치하는지 물어보았을 때 맞다는 회신을 받았습니다.

![image](https://user-images.githubusercontent.com/14256139/210080949-f2eadf48-a445-4a1a-8ee4-2d3a06d1e292.png)

지금 조직 내에서 토큰화해서 관리하는 컬러, 타이포그래피 그리고 자주 변경되는 마진 정도의 요소 등의 우선순위로 시각적 변화가 수월하게 반영되길 바라는 니즈를 확인하였습니다. 그래서 우선 시각적 요소의 업데이트를 쉽게 해주는 것을 목표로 MVP를 구현하고 다시 인터뷰를 갖기로 하고 인터뷰를 마쳤습니다.

인터뷰 때 인터뷰이에겐 언급하지 않았지만 최근에 페이스북에서 접했던 아래의 글을 떠올리면서 질문을 했었습니다. 시각적 요소 또한 비용에 대한 걱정없이 시도할 수 있게 하는게 프로덕트에 매우 유의미할 수 있다고 생각해보았습니다.

> 끊임없이 자동화해야합니다.
자동화는 10분 걸리는 일을 1초로 바꿉니다. 이러한 노력의 댓가는 “시간을 10분 아낀다” 정도로 가볍지 않습니다. 업무를 자동화 함에 따라 얻는 댓가는 “10분 동안 더 깊고 생산적인 고민을 한다”입니다.
예를 들어, A팀은 API 요청과 응답의 JSON 인터페이스를 서버와 클라이언트 개발자가 일일히 손으로 맞춘다고 가정해봅시다. A팀에서 인터페이스를 바꾸는 일은 동료에게 리소스를 요구하는 것이기 때문에 소통 비용이 높습니다. 따라서 혹시 인터페이스에 문제가 있더라도 누구든 함부로 기존 인터페이스를 바꾸자는 이야기를 꺼내지 못합니다. (비가역적)
반대로, B팀은 스키마*를 통해 서버, 클라이언트 코드를 자동 생성한다고 합시다. 이 팀은 인터페이스를 바꾸는 것은 CLI 호출 1초면 됩니다. B팀은 인터페이스 변경이 너무나도 쉽기 때문에, 소통 비용에 대한 걱정 없이 더 많은 시도를 할 수 있습니다. (“이렇게 바꿔보면 어떨까요?”라는 질문을 신규입사자도 꺼낼 수 있습니다) 결론적으로 인터페이스 자체에 대한 더 깊은 고민을 할 수 있게 됩니다. (가역적)
A팀과 B팀은 당장은 10분의 차이만 있습니다. 하지만 결과적으로 B팀은 더 지속가능하고 확장성있는 API 설계를 통해 10분보다 훨씬 더 많은 시간을 아꼈을 것입니다.
*스키마: OpenAPI Specification(JSON Schema), GraphQL, Protobuf 등

**Reply by peanudge** [reply]
인터뷰 해주셔서 감사합니다. 👍 디자이너들이 디자인된 컴포넌트가 어디서 사용되는지 쉽게 찾고 싶어하는 니즈는 확인한 것 같아요.

**Comment by ppamppamman** [comment]
220107 온보딩 회의
- webpack 플러그인과 관련하여 서로의 이해도를 짚어보는 회의를 짧게 가졌습니다.
- 이로부터 도출된 액션 아이템은 아래와 같습니다.

### Action Item

- [ ]  edge-case 판단 - 현재 작업중
- [ ]  스토리북 어떻게 데이터를 밀어넣을 것인지.
    - 스토리북을 볼 때의 개발자의 관심사 / 디자이너의 관심사가 무엇일지.
      - [아래의 리플라이](https://github.com/InnerCircleA/design-system/discussions/12#discussioncomment-4624247)에 정리
- [ ]  props의 변화 양상
    - 어떻게 쓰이고 있는지 → props별로 어떤 value가 쓰이고 있는지.
- [ ]  lint로 어노테이션에 대한 휴먼에러 방지
    - filename, func name

**Comment by ppamppamman** [comment]
디자인 QA에 대해 디자인시스템의 활용으로 일부 대체할 수 있을지에 대한 논의
> 팀 디자이너분과의 인터뷰를 토대로 작성되었습니다.

- 디자인 QA의 정의
  - figma의 작업본이 올바르게 적용되었는지 빌드된 FE개발자의 작업물을 디자이너가 크롬으로 확인하는 행위 
- 디자이너의 어려움과 니즈
  - 어려움인 줄 알았으나 아니었던 점
    - 작업 대상이 되는 페이지에 파운데이션 요소가 올바르게 사용되었는지를 파악하는 것은 의외로 큰 어려움이 아님. 예를 들면, border에 대한 px 차이 등을 잡아내는 것이 어렵지 않고, 디자이너분들이 공통적으로 그리 어렵지 않게 파악할 수 있으나, 개발자는 그러지 못하다는 것이 허들로 존재.
  - 어려운 점
    - 디자인 QA에 따르는 커뮤니케이션을 위한 문서작업이나 미팅을 한 번 이상 열어야 하는 것이 어려운 상황. 오히려 디자인적으로 업무하는 리소스보다 부가적인 리소스가 더 크게 들어가고 있는 셈.
  - 니즈
    - 만약 개발자가 개발선에서 디자인시스템의 컴포넌트가 적용되지 못하거나, 지시되었던 파운데이션 요소가 적용되지 않은 상황을 일찍 알아챌 수 있다면, 디자인 QA에 대한 업무량이 본질적으로 줄어들 수 있을 것.
    - 혹은 디자이너가 미리 사용된 컴포넌트에 대한 리포트를 받아볼 수 있으면 그것을 기반으로 커뮤니케이션을 할 수 있을 것이라 판단.
- 아이디어
  - 앞서 도출되었던 action item 중 각 컴포넌트의 props가 어떻게 쓰이고 있는지를 파악하고자 하는 action item이 있었는데요, 이 부분이 특정 단위별로 lint report와 비슷한 모습으로 확인될 수 있다면, 페이지별 사용 리포트를 통해 visual regression test 이하의 레벨에서 확인해볼 수 있을 것이라 판단.
  ```js
    // 기존
    {
      "project": "poc@1.0.0",
      "name": "MainPage.ts",
      "alias": "메인 페이지",
      "importNames": ["ConfirmButton", "CancelButton"]
    }

    // 개선
    {
      "project": "poc@1.0.0",
      "components": [
        {
            "name" : "ConfirmButton",
            "import" : "MainPage.ts",
            "props" : {
              "value": "보내기"
              "color" : "#nnnnnn" | "Semantic.action.primary", 
              ...
             },
        },
        {
            "name" : "ConfirmButton",
            "import" : "MainPage.ts",
            "props" : {
              "value": "확인하기"
              "color" : "#mmmmmm" | "Semantic.action.secondary", 
              ...
             },
        },

      ]
    }

  ```

- 찾아본 점
  - https://www.chromatic.com/docs/figma 
  - https://dev.to/reallifenero/how-to-implement-chromatic-plugin-x-figma-addon-for-vue-and-storybook-1mf6

**Reply by ppamppamman** [reply]
오 근데 작성하고 다시 읽어보니 앞서 규태님이 고민하셨던 내용과 비슷한 것인가? 하는 생각이 드네요..? 제가 완벽하게 트래킹하고 있지 못한 상황이라 비슷한 느낌을 받았습니다.

**Reply by peanudge** [reply]
> 지시되었던 파운데이션 요소가 적용되지 않은 상황

여기서 어떤 파운데이션 요소들이었는지 구체적으로 알 수 있을까요?

**Comment by ppamppamman** [comment]
```
export interface ButtonProps extends Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type' | 'css'>; {
  children: React.ReactNode;
  htmlType?: 'submit' | 'button' | 'reset';
  type?: 'primary' | 'secondary' | 'text' | 'link';
  size?: 'large' | 'medium' | 'small';
  loading?: boolean;
  icon?: JSX.Element;
}
```

여기서 크기를 의미하는 값인 'large' | 'medium' | 'small'는 미리 정의된 크기, 마진, 패딩과 관련된 파운데이션 토큰값의 세트를 map자료형 key의 역할을 하고 type의 경우는 color에 대해 미리 정의해놓은 파운데이션 토큰의 집합으로부터 만들어진 map 자료형의 key 역할을 할 수 있습니다.

예를 들면 primary 같은 경우는 이런느낌으로 다뤄볼 수 있을 것 같아요.
```
const primary = css`
  background-color: ${SemanticColor.Action.Primary.Normal};
  color: ${PaletteColor.white};
  ${TypographyStyles.Body2}

  @media (hover: hover) {
    &:enabled:hover {
      background-color: ${SemanticColor.Action.Primary.Hover};
    }
  }

  &:enabled:active {
    box-shadow: 0 0 0 3px ${PaletteColor.blue['100']};
    transition: 0s;
  }

  &:disabled {
    background-color: ${SemanticColor.Action.Disabled};
    color: ${PaletteColor.white};
  }
`;
```

이 구조는 미리 정의한 값에 의존하고 있는데요, 각 스타일 토큰을 따로 받아 오버라이딩 하는 경우도 고려해볼 수 있을 것 같습니다.
